### PR TITLE
Fix Resize shape inference precision for scaled output dimensions

### LINKED ISSUE #4919: [Resize] which conversion mode should be taken in RESIZE output_dim computation, Floor or Round?
# Ask a Question

output_dim computation is not unified on how to convert float to int.
Is this a bug?

### Question
In Resize-18, `keep_aspect_ratio_policy ` has three different options, "stretch","not_larger","not_smaller".

We have round mode when `policy` is not the default from the spec:
`out_size[d] = round_int(scale * in_size[i])`

However, for the default policy, `stretch`,  we have floor instead to calculate the output shape.

The question is, Python.float is not mapping to cpp.float. so we get a shape different eventually.

### Further information

One of the cases where we will run into the output-shape in-consistency.

Say we want to do a Resize operation for image which has shape (224,224), then enlarge it into (706,706), and we want to resize back to (224,224) in the second step. Apprently, if we want  to make this work and trying to give it a scale, (224/706,224/706). Then things are in a mess in C++, because 224*(224/706)=223.99999999, if it's truncated into INT with floor, we got 223, which is not expected as we want.

Python does work for this case, since python.float is same as cpp.double.

### PR DESCRIPTION
**Pull request by elwhyjay**
### Motivation and Context

Fixes #4919.

The two `Resize` shape inference helpers in `onnx/defs/tensor/utils.cc` multiply the input dimension by the scale through a `float32` intermediate. `float32` has a 24-bit mantissa, so any input dimension `> 2**24` is silently
truncated *before* `floor` runs. With `input dim = 16777217` and `scales = 1.0`, the inferred output becomes `16777216` instead of `16777217`:
```cpp
    static_cast<int64_t>(std::floor(static_cast<float>(16777217) * 1.0f))
    // -> floor(16777216.0f * 1.0f) -> 16777216
```
The 1-off result is not a `floor` rounding decision — it happens at the `static_cast<float>` step. Maintainers previously decided to keep `floor` to align with PyTorch `F.interpolate`, so this PR does **not** change the rounding mode. It only widens the intermediate to `double` so `floor` operates on the precise value.

### What changed

`onnx/defs/tensor/utils.cc` — both helpers updated identically:

- `resizeShapeInferenceHelper` (current-version)
- `resizeShapeInferenceHelper_opset7_to_10` (legacy)

```cpp
// before
static_cast<int64_t>(std::floor(
    static_cast<float>(input_shape.dim(i).dim_value()) * scales_data[i]));

// after
static_cast<int64_t>(std::floor(
    static_cast<double>(input_shape.dim(i).dim_value())
    * static_cast<double>(scales_data[i])));
```
Both are reachable depending on the model's opset, so both must change for the fix to take effect end-to-end.

### Out of scope

`KeepAspectRatioHelper` at `utils.cc:56` uses a similar `static_cast<float>` pattern but on the `sizes`-input path with `roundf`, which has different semantics. Intentionally left out of this PR to keep the change focused — happy to follow up separately if maintainers want that path addressed.

### Tests

Three regression tests added to `onnx/test/shape_inference_test.py`, parameterized over all Resize opsets:

- `test_resize_scale_precision_large_dim` — current-version helper, `scale=1`
- `test_resize_scale_precision_large_dim_non_unit_scale` — `scale=2.0` path  (expected `33554434`, float32 produces `33554432`)
- `test_resize_scale_precision_large_dim_opset10` — opset 7-10 helper

All three fail on `main` and pass with this change. Full `shape_inference_test.py` continues to pass with no other regressions.

### CODE REVIEW

File: onnx/test/shape_inference_test.py [hunk 1]
```diff
@@ -1325,6 +1324,75 @@ def test_resize_scale_raw_data(self, _, version) -> None:
             opset_imports=[helper.make_opsetid(ONNX_DOMAIN, version)],
         )
 
+    @parameterized.expand(all_versions_for("Resize"))
+    def test_resize_scale_precision_large_dim(self, _, version) -> None:
```
Comment (cbourjau) [inline onnx/test/shape_inference_test.py]: Could you remove this code duplication by using helper function or test parametrization?

**Review by cbourjau — CHANGES_REQUESTED** [review]
Thanks for the fix. I have I minor comment

File: onnx/test/shape_inference_test.py (hunk 1 above)
Comment (elwhyjay) [reply onnx/test/shape_inference_test.py]: Thanks! Extracted into _check_resize_scale_precision. Each test now just declares skip condition, scales, and expected dim. PTAL

**Comment by elwhyjay** [comment]
Friendly ping @cbourjau — refactor applied in 2246f221. PTAL when you have a moment, happy to address
any further feedback.

**Comment by cbourjau** [comment]
Thanks, I shrank the test case a bit further. My gosh, `parameterized` is a pain to use :D

**Comment by elwhyjay** [comment]
Thanks a lot @cbourjau ! And yes, `parameterized` can be surprisingly tricky.   Your simplification looks much cleaner. I appreciate the cleanup 👍

**Review by cbourjau — APPROVED** [review]

**Review by justinchuby — APPROVED** [review]
Thanks!

**Review by fdwr — COMMENTED** [review]
Thanks Lee. The test case value [looks familiar](https://github.com/onnx/onnx/issues/4919#issuecomment-1782207470) :b.
